packages/fast-usdc/test/fast-usdc.contract.test.ts

```diff
@@ -510,115 +542,460 @@ test('LP repay', async t => {
+    racer2: makeCustomer('Racer2', cctp, txPub.publisher, feeConfig),
+    racer3: makeCustomer('Racer3', cctp, txPub.publisher, feeConfig),
+  };
+
+  await cust.racer3.checkPoolAvailable(t, 125_000_000n, metricsSub);
+
+  const bridgePos = snapshot();
+  const [sent1, sent2, sent3] = await Promise.all([
+    cust.racer1.send(t, 110_000_000n, 'osmo1234a'),
+    cust.racer2.send(t, 120_000_000n, 'osmo1234b'),
+    cust.racer3.send(t, 125_000_000n, 'osmo1234c'),
+  ]);
+  cust.racer1.checkSent(t, since(bridgePos));
+  cust.racer2.checkSent(t, since(bridgePos));
+  // TODO/WIP: cust.racer3.checkSent(t, since(bridgePos), 'forward - LP depleted');
   await transmitTransferAck();
-  // Nothing we can check here, unless we want to inspect calls to `trace`.
-  // `test/exos/advancer.test.ts` covers calls to `log: LogFn` with mocks.
-  // This is still helpful to call, so we can observe "Advance transfer
-  // fulfilled" in the test output.
+  await transmitTransferAck();
+  await transmitTransferAck();
+  await Promise.all([mint(sent1), mint(sent2), mint(sent3)]);
+});
+
+test.serial('STORY05(cont): LPs withdraw all liquidity', async t => {
+  const {
+    sync,
+    common: {
+      brands: { usdc },
+    },
+  } = t.context;
+
+  const lp = await sync.lp.promise;
+  const [a, b] = await Promise.all([
+    E(lp.lp200).withdraw(t, 1),
+    E(lp.lp50).withdraw(t, 1),
+  ]);
+  t.log({ a, b, sum: add(a, b) });
+  t.truthy(a);
+  t.truthy(b);
+});
+
+test.serial('STORY09: insufficient liquidity: no FastUSDC option', async t => {
+  // STORY09 - As the Fast USDC end user,
+  // I should see the option to use Fast USDC unavailable
+  // on the UI (and unusable) if there are not funds in the
+  // MarketMaker’s account
+  const {
+    common: {
+      commonPrivateArgs: { feeConfig },
+    },
+    evm: { cctp, txPub },
+    startKit: { metricsSub },
+  } = t.context;
+  const early = makeCustomer('Unice', cctp, txPub.publisher, feeConfig);
+  const available = await early.checkPoolAvailable(t, 5_000_000n, metricsSub);
+  t.false(available);
+});
+
+test.serial('C20 - Contract MUST function with an empty pool', async t => {
+  const {
+    common: {
+      commonPrivateArgs: { feeConfig },
+      utils: { transmitTransferAck },
+    },
+    evm: { cctp, txPub },
+    startKit: { metricsSub },
+    bridges: { snapshot, since },
+    mint,
+  } = t.context;
+  const custEmpty = makeCustomer('Earl', cctp, txPub.publisher, feeConfig);
+  const bridgePos = snapshot();
+  const sent = await custEmpty.send(t, 150_000_000n, 'osmo123');
+  const bridgeTraffic = since(bridgePos);
+  await mint(sent);
+  custEmpty.checkSent(t, bridgeTraffic, 'forward');
+  t.log('No advancement, just settlement');
+  await transmitTransferAck(); // ack IBC transfer for forward
 });
+
+// advancedEarly stuff
+test.todo(
+  'C12 - Contract MUST only pay back the Pool only if they started the advance before USDC is minted',
```

Comment:
This is advance early, but also could be the same scenario about. We could inspect the outgoing LCA bridge messages to confirm they sent the requested amount and not the net of fees amount

Reply:
> We could inspect the outgoing LCA bridge messages to confirm they sent the requested amount and not the net of fees amount

`checkSent` with `'forward'` does check that.

But we didn't test the `advancedEarly` case, i.e. the case where the mint comes in between ADVANCING and ADVANCED.